src/com/google/javascript/jscomp/ConvertChunksToESModules.java

```diff
@@ -0,0 +1,297 @@
+ * exports for them.
+ *
+ * chunk1.js
+ * <pre>var a = 1; function b() { return a }</pre>
+ *
+ * chunk2.js
+ * <pre>console.log(a);</pre>
+ *
+ * becomes
+ *
+ * chunk1.js
+ * <pre>var a = 1; var b = function b() { return a }; export {a};</pre>
+ *
+ * chunk2.js
+ * <pre>import {a} from './chunk1.js'; console.log(a);</pre>
+ *
+ * This allows splitting code into es modules that depend on each other's symbols, without using
+ * a global namespace or polluting the global scope.
+ */
+final class ConvertChunksToESModules implements CompilerPass {
+  private final AbstractCompiler compiler;
+  private final Map<JSModule, Set<String>> crossChunkExports = new HashMap<>();
+  private final Map<JSModule, Map<JSModule, Set<String>>> crossChunkImports = new HashMap<>();
+
+  static final DiagnosticType ASSIGNMENT_TO_IMPORT = DiagnosticType.error(
+      "JSC_IMPORT_ASSIGN",
+      "Imported symbol \"{0}\" in chunk \"{1}\" cannot be assigned");
+
+  static final DiagnosticType UNABLE_TO_COMPUTE_RELATIVE_PATH = DiagnosticType.error(
+      "JSC_UNABLE_TO_COMPUTE_RELATIVE_PATH",
+      "Unable to compute relative import path from \"{0}\" to \"{1}\"");
+
+  /**
+   * Constructor for the ConvertChunksToESModules compiler pass.
+   *
+   * @param compiler The JSCompiler, for reporting code changes.
+   */
+  ConvertChunksToESModules(AbstractCompiler compiler) {
+    this.compiler = compiler;
+  }
+
+  @Override
+  public void process(Node externs, Node root) {
+    // Find global names that are used in more than one chunk. Those that
+    // are have to have import and export statements added.
+    NodeTraversal.traverse(compiler, root, new FindCrossChunkReferences());
+
+    // Force every output chunk to parse as an ES Module. If a chunk has no imports and
+    // no exports, add an empty export list to generate an empty export statement:
+    // example: export {};
+    for (JSModule chunk : compiler.getModuleGraph().getAllModules()) {
+      if (!crossChunkExports.containsKey(chunk) &&
+          !crossChunkImports.containsKey(chunk) &&
+          chunk.getInputs().size() > 0) {
+        crossChunkExports.put(chunk, new HashSet<>());
+      }
+    }
+
+    ConvertChunkSourcesToModules();
+    AddExportStatements();
+    AddImportStatements();
+  }
+
+  /**
+   * Move all code in a chunk into the first input and mark it as an ESModule.
+   * At this point in the compilation, all input files should be scripts.
+   */
+  private void ConvertChunkSourcesToModules() {
+    for (JSModule chunk : compiler.getModuleGraph().getAllModules()) {
+      if (chunk.getInputs().size() == 0) {
+        continue;
+      }
+
+      CompilerInput firstInput = null;
+      for (CompilerInput input : chunk.getInputs()) {
+        Node astRoot = input.getAstRoot(compiler);
+        FeatureSet scriptFeatures = NodeUtil.getFeatureSetOfScript(astRoot);
+        checkState(!scriptFeatures.contains(FeatureSet.ES6_MODULES));
+        if (firstInput == null) {
```

Comment:
Could you add comments near here to show what you intend the AST structure to look like when this is done?

Maybe:

```java
// Original AST:
// [ROOT] top JS sources root
//     [SCRIPT] // 1st chunk's 1st input
//         statements from 1st chunk's 1st script
//     [SCRIPT] // 1st chunk's 2nd input
//         statements from 1st chunk's 2nd script
//     ...
//     [SCRIPT] // 1st chunk's nth input
//     [SCRIPT] // 2nd chunk's 1st input
//     ...
//
// After this method
// [ROOT] top JS sources root
//     [SCRIPT] // 1st chunk's 1st input script
//         [MODULE_BODY] // new module body to contain the 1st chunk
//             // statements from all scripts in the first chunk
//             // All script nodes from the other inputs of the chunk are now empty
//     [SCRIPT] // 2nd chunk's 1st input script
//         [MODULE_BODY] // new module body to contain the 2nd chunk
//             // statements from all scripts in the first chunk
//             // All script nodes from the other inputs of the chunk are now empty
```

Reply:
done